### PR TITLE
feat(lwgenerate): Add initial hcl generation support

### PR DESCRIPTION
**Pull request by ipcrm**
## Summary

Add initial support for Terraform code generation

## How did you test this change?

See [examples](https://github.com/ipcrm/go-sdk/tree/ipcrm/ally-392/lwgenerate/lwgenerate/_examples) & [tests](https://github.com/ipcrm/go-sdk/tree/ipcrm/ally-392/lwgenerate/lwgenerate)

## Issue

https://lacework.atlassian.net/browse/ALLY-392

### CODE REVIEW

**Comment by afiune** [comment]
Make it so!

File: lwgenerate/hcl.go [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+				if err != nil {
+					return nil, err
+				}
+				data[attrKey] = value
+			}
+			block.Body().SetAttributeValue(key, cty.ObjectVal(data))
+		default:
+			return nil, errors.New("Unknown type")
+		}
+	}
```
Comment (dmurray-lacework) [inline lwgenerate/hcl.go:123]: This block is a little difficult to read, perhaps extracting it to it's own function with a clear name and comments would help.

File: lwgenerate/aws.go [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+	// Existing IAM Role Name
+	Name string
+
+	// Existing IAM Role External Id
+	ExternalId string
+}
+
+type GenerateAwsTfConfigurationArgs struct {
+	// Should we configure cloudtrail integration in LW?
+	ConfigureCloudtrail bool
```
Comment (dmurray-lacework) [inline lwgenerate/aws.go:23]: Do all these fields need to be public?

File: lwgenerate/aws.go [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+
+	// Should we force destroy the bucket if it has stuff in it? (only relevant on new Cloudtrail creation)
+	ForceDestroyS3Bucket bool
+
+	// For AWS Subaccounts in consolidated CT setups
+	// TODO @ipcrm future: what about many individual ct/config integrations together?
+	Profiles map[string]string
+
+	// Lacework Profile to use
+	LaceworkProfile string
```
Comment (dmurray-lacework) [inline lwgenerate/aws.go:55]: Do all these fields need to be public?

File: lwgenerate/aws.go [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+
+type ExistingIamRoleDetails struct {
+	// Existing IAM Role ARN
+	Arn string
+
+	// Existing IAM Role Name
+	Name string
+
+	// Existing IAM Role External Id
+	ExternalId string
```
Comment (dmurray-lacework) [inline lwgenerate/aws.go:18]: Are these fields required to be exported?

Nit: Also perhaps not all fields require comments. I don't believe we follow this pattern elsewhere.

File: lwgenerate/_examples/aws/main.go [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+func main() {
+	hcl, err := lwgenerate.GenerateAwsTerraformCode(&lwgenerate.GenerateAwsTfConfigurationArgs{
+		ConfigureCloudtrail: true,
+		ConfigureConfig:     true,
+		AwsRegion:           "us-east-1",
+	})
+
+	if err != nil {
+		fmt.Println(err.Error())
```
[the remark this answers is not in the document]
Comment (ipcrm) [reply lwgenerate/_examples/aws/main.go:17]: 👍

File: lwgenerate/_examples/hcl/main.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+func exampleModule() {
+	data, err := lwgenerate.CreateModule(&lwgenerate.HclModule{
+		Name:   "servers",
+		Source: "./app-cluster",
+		Attributes: map[string]interface{}{
+			"servers": 5,
+		}})
+
+	if err != nil {
+		fmt.Println(err.Error())
```
[the remark this answers is not in the document]
Comment (ipcrm) [reply lwgenerate/_examples/hcl/main.go:18]: 👍

File: lwgenerate/aws.go [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+
+	// Lacework Profile to use
+	LaceworkProfile string
+}
+
+// Ensure all combinations of inputs our valid for supported spec
+func validateArguments(args *GenerateAwsTfConfigurationArgs) error {
+	// Validate one of config or cloudtrail was enabled; otherwise error out
+	if !args.ConfigureCloudtrail && !args.ConfigureConfig {
+		return errors.New("must enable cloudtrail or config integration")
```
[the remark this answers is not in the document]
Comment (ipcrm) [reply lwgenerate/aws.go:62]: 👍

File: lwgenerate/hcl.go [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+func convertTypeToCty(value interface{}) (cty.Value, error) {
+	switch v := value.(type) {
+	case string:
+		return cty.StringVal(v), nil
+	case int:
+		return cty.NumberIntVal(int64(v)), nil
+	case bool:
+		return cty.BoolVal(v), nil
+	default:
+		return cty.NilVal, errors.New("Unknown attribute value type")
```
[the remark this answers is not in the document]
Comment (ipcrm) [reply lwgenerate/hcl.go:60]: 👍

File: lwgenerate/aws.go [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+	// For AWS Subaccounts in consolidated CT setups
+	// TODO @ipcrm future: what about many individual ct/config integrations together?
+	Profiles map[string]string
+
+	// Lacework Profile to use
+	LaceworkProfile string
+}
+
+// Ensure all combinations of inputs our valid for supported spec
+func validateArguments(args *GenerateAwsTfConfigurationArgs) error {
```
[the remark this answers is not in the document]
Comment (ipcrm) [reply lwgenerate/aws.go:59]: 👍

File: lwgenerate/aws.go [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+			args.ExistingIamRole.ExternalId == "" {
+			return errors.New("when using an existing IAM role, existing role ARN, name, and external ID all must be set")
+		}
+	}
+
+	return nil
+}
+
+// Generate new Terraform code based on the supplied args.
+func GenerateAwsTerraformCode(args *GenerateAwsTfConfigurationArgs) (string, error) {
```
[the remark this answers is not in the document]
Comment (ipcrm) [reply lwgenerate/aws.go:83]: Ya, I don't think there is a reason for this now.  Originally I was storing the blocks in the struct and had to modify it, but now no reason to.

File: lwgenerate/hcl.go (hunk 1 above)
Comment (ipcrm) [reply lwgenerate/hcl.go:123]: 👍

File: lwgenerate/aws.go (hunk 3 above)
Comment (ipcrm) [reply lwgenerate/aws.go:55]: They'll have to be unless it would be better to create a constructor with getter/setters for all of these

File: lwgenerate/aws.go (hunk 4 above)
Comment (ipcrm) [reply lwgenerate/aws.go:18]: Similar to above, these get accessed outside lwgenerate package.

Comment (ipcrm) [reply lwgenerate/aws.go:18]: Re doc strings; I can remove these

Comment (afiune) [reply lwgenerate/aws.go:18]: I disagree with @dmurray-lacework nitpick, comments are great and we should keep and encourage them.

Example that we do this here: https://github.com/lacework/go-sdk/blob/main/cli/cmd/compliance.go#L35-L60

File: lwgenerate/aws.go (hunk 3 above)
Comment (ipcrm) [reply lwgenerate/aws.go]: No longer public in the rework

File: lwgenerate/aws.go [hunk 11]
```diff
@@ -0,0 +1,381 @@
...
+}
+
+// Set the AWS Profile to utilize for the main AWS provider
+func AwsTerraformWithAwsProfile(name string) AwsTerraformModifier {
+	return func(c *GenerateAwsTfConfigurationArgs) {
+		c.awsProfile = name
+	}
+}
+
+// Set the Laceowork Profile to utilize when integrating
```
Comment (dmurray-lacework) [inline lwgenerate/aws.go:133]:
```suggestion
// Set the Lacework Profile to utilize when integrating
```

File: lwgenerate/aws.go [hunk 12]
```diff
@@ -0,0 +1,381 @@
...
+func AwsTerraformWithLaceworkProfile(name string) AwsTerraformModifier {
```
Comment (dmurray-lacework) [inline lwgenerate/aws.go:134]:
```suggestion
func AwsTerraformWithLaceworkProfile(profile string) AwsTerraformModifier {
	return func(c *GenerateAwsTfConfigurationArgs) {
		c.laceworkProfile = profile
	}```

File: lwgenerate/aws.go (hunk 11 above)
Comment (ipcrm) [reply lwgenerate/aws.go:133]: 👍

File: lwgenerate/_examples/aws/main.go [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/lacework/go-sdk/lwgenerate"
+)
+
+func main() {
+	hcl, err := lwgenerate.NewAwsTerraform("us-east-1", true, true).Generate()
```
Comment (dmurray-lacework) [inline lwgenerate/_examples/aws/main.go:11]: Can we add more examples with some of the more complicated use cases and comments.

**Comment by afiune** [comment]
make it so

File: lwgenerate/hcl.go [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+}
+
+// Convert blocks to a string
+func CreateHclStringOutput(blocks []*hclwrite.Block) string {
+	file := hclwrite.NewEmptyFile()
+	body := file.Body()
+
+	for _, b := range blocks {
+		body.AppendBlock(b)
+		body.AppendNewline()
```
Comment (afiune) [inline lwgenerate/hcl.go:297]: Add a test for this please, what if a block has a `nil` pointer? We should always check
```suggestion
		if b != nil {
			body.AppendBlock(b)
			body.AppendNewline()
		}
```

File: lwgenerate/hcl.go [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+			if len(v) > 0 {
+				blocks = append(blocks, v...)
+			}
+		default:
+			return nil, errors.New("Unknown type supplied!")
+		}
+	}
+
+	return blocks, nil
+}
```
Comment (afiune) [inline lwgenerate/hcl.go:366]: As much as I like the fact that you are helping the user combine blocks, I don't think we should throw an error here, but instead, only add the two cases you have for a single pointer and an array of pointers. 

```suggestion
func CombineHclBlocks(results ...interface{}) []*hclwrite.Block {
	blocks := []*hclwrite.Block{}
	// Combine all blocks into single flat slice
	for _, result := range results {
		switch v := result.(type) {
		case *hclwrite.Block:
			if v != nil {
				blocks = append(blocks, v)
			}
		case []*hclwrite.Block:
			if len(v) > 0 {
				blocks = append(blocks, v...)
			}
		default:
			continue
		}
	}

	return blocks
}
```

File: lwgenerate/hcl.go [hunk 16]
```diff
@@ -0,0 +1,366 @@
...
+// Convert blocks to a string
+func CreateHclStringOutput(blocks []*hclwrite.Block) string {
+	file := hclwrite.NewEmptyFile()
+	body := file.Body()
+
+	for _, b := range blocks {
+		body.AppendBlock(b)
+		body.AppendNewline()
+	}
+	return string(file.Bytes())
```
Comment (afiune) [inline lwgenerate/hcl.go:299]: For some reason there are trailing spaces and carrier returns at the end, this trims them all
```suggestion
	return strings.TrimRight(string(file.Bytes()), "\n")
```

**Review by afiune — CHANGES_REQUESTED** [review]
tenor-55382657

File: lwgenerate/hcl.go (hunk 14 above)
Comment (ipcrm) [reply lwgenerate/hcl.go:297]: 👍

File: lwgenerate/hcl.go (hunk 15 above)
Comment (ipcrm) [reply lwgenerate/hcl.go:366]: 👍

File: lwgenerate/hcl.go (hunk 16 above)
Comment (ipcrm) [reply lwgenerate/hcl.go:299]: 👍

Comment (ipcrm) [reply lwgenerate/hcl.go:299]: Two things were causing this 

1) I added new lines to each block to provide proper spacing (L297 above)
2) The examples were using `fmt.Println` which added yet another space (HCL examples are printing single blocks so each block would have this behavior giving double spacing vs AWS which is one slice of blocks so it only had 1 trailing space)

I've resolved both of these by fixing the examples and by only appending newlines to non-last blocks.

File: lwgenerate/_examples/aws/main.go (hunk 13 above)
Comment (ipcrm) [reply lwgenerate/_examples/aws/main.go]: 👍

File: lwgenerate/aws.go [hunk 17]
```diff
@@ -0,0 +1,373 @@
...
+		}
+	}
+
+	return nil
+}
+
+type AwsTerraformModifier func(c *GenerateAwsTfConfigurationArgs)
+
+// Create new AWS Terraform configuration
+func NewAwsTerraform(region string, enableConfig bool, enableCloudtrail bool, mods ...AwsTerraformModifier) *GenerateAwsTfConfigurationArgs {
```
[the remark this answers is not in the document]
Comment (ipcrm) [reply lwgenerate/aws.go]: 👍

File: lwgenerate/aws/aws.go [hunk 18]
```diff
@@ -0,0 +1,383 @@
...
+
+func createConfig(args *GenerateAwsTfConfigurationArgs) ([]*hclwrite.Block, error) {
+	source := "lacework/config/aws"
+	version := "~> 0.1"
+
+	blocks := []*hclwrite.Block{}
+	if args.Config {
+		moduleDetails := []lwgenerate.HclModuleModifier{}
+		// Add main account
+		// block := NewModule("aws_config", source, HclModuleWithVersion(version))
```
Comment (afiune) [inline lwgenerate/aws/aws.go:301]: do we need this?

File: lwgenerate/aws/aws.go [hunk 19]
```diff
@@ -0,0 +1,383 @@
...
+			cloudTrailModule),
+	)
+	return hclBlocks, nil
+}
+
+func createRequiredProviders() (*hclwrite.Block, error) {
+	return lwgenerate.CreateRequiredProviders(
+		lwgenerate.NewRequiredProvider("lacework",
+			lwgenerate.HclRequiredProviderWithSource("lacework/lacework"),
+			lwgenerate.HclRequiredProviderWithVersion("~> 0.3")))
```
Comment (afiune) [inline lwgenerate/aws/aws.go:237]: We are already in [version `0.12.X` ](https://registry.terraform.io/providers/lacework/lacework/latest)- let us bump it already :)
```suggestion
			lwgenerate.HclRequiredProviderWithVersion("~> 0.12")))
```

File: lwgenerate/hcl.go (hunk 16 above)
Comment (afiune) [reply lwgenerate/hcl.go:306]: 👌🏽

**Review by afiune — APPROVED** [review]
tenor-219999923

File: lwgenerate/aws/aws.go [hunk 20]
```diff
@@ -0,0 +1,383 @@
...
+// Note: Additional configuration details may be set using modifiers of the AwsTerraformModifier type
+//
+// Basic usage: Initialize a new AwsTerraformModifier struct, with a non-default AWS profile set. Then use generate to
+//              create a string output of the required HCL.
+//
+//   hcl, err := aws.NewAwsTerraform(
+//     "us-east-1",
+//     true,
+//     true,
+//     aws.WithAwsProfile("mycorp-profile")).Generate()
```
Comment (afiune) [inline lwgenerate/aws/aws.go:122]: nit: no need to split this lines that much
```suggestion
//   hcl, err := aws.NewAwsTerraform("us-east-1", true, true,
//     aws.WithAwsProfile("mycorp-profile")).Generate()
```

File: lwgenerate/aws/aws.go (hunk 18 above)
Comment (ipcrm) [reply lwgenerate/aws/aws.go:301]: no

File: lwgenerate/aws/aws.go (hunk 19 above)
Comment (ipcrm) [reply lwgenerate/aws/aws.go:237]: 👍

File: lwgenerate/aws/aws.go (hunk 20 above)
Comment (ipcrm) [reply lwgenerate/aws/aws.go:122]: 👍

File: lwgenerate/aws/aws.go [hunk 21]
```diff
@@ -0,0 +1,380 @@
...
+//
+// Note: Additional configuration details may be set using modifiers of the AwsTerraformModifier type
+//
+// Basic usage: Initialize a new AwsTerraformModifier struct, with a non-default AWS profile set. Then use generate to
+//              create a string output of the required HCL.
+//
+//   hcl, err := aws.NewAwsTerraform("us-east-1", true, true,
+//     aws.WithAwsProfile("mycorp-profile")).Generate()
+//
+func NewAwsTerraform(region string, enableConfig bool, enableCloudtrail bool, mods ...AwsTerraformModifier) *GenerateAwsTfConfigurationArgs {
```
Comment (dmurray-lacework) [inline lwgenerate/aws/aws.go:121]: 1 final change request. Now that there is a separate package for aws, can you omit Aws from the name of this func.
aws.NewTerraform(...)
vs
aws.NewAwsTerraform(..)

**Review by dmurray-lacework — APPROVED** [review]

File: lwgenerate/aws/aws.go (hunk 21 above)
Comment (ipcrm) [reply lwgenerate/aws/aws.go]: No problem; updated!

**Comment by afiune** [comment]
make it so
